README.md
# Get URI META

[![Latest Version on Packagist](https://img.shields.io/packagist/v/chuoke/uri-meta.svg?style=flat-square)](https://packagist.org/packages/chuoke/uri-meta)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/chuoke/uri-meta/run-tests?label=tests)](https://github.com/chuoke/uri-meta/actions?query=workflow%3ATests+branch%3Amaster)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/chuoke/uri-meta/Check%20&%20fix%20styling?label=code%20style)](https://github.com/chuoke/uri-meta/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amaster)
[![Total Downloads](https://img.shields.io/packagist/dt/chuoke/uri-meta.svg?style=flat-square)](https://packagist.org/packages/chuoke/uri-meta)

---

This package can be used as to extract the META information of the URI.

## Installation

You can install the package via composer:

```bash
composer require chuoke/uri-meta
```

## Usage

```php
    $extracter = new UriMetaExtracter();

    $uriMeta = $extracter->extract('https://www.php.net/');

    echo $uriMeta->title(); // PHP: Hypertext Preprocessor

    print_r($uriMeta->toArray());
    // Array
    // (
    //     [uri] => https://www.php.net/
    //     [host] => www.php.net
    //     [scheme] => https
    //     [title] => PHP: Hypertext Preprocessor
    //     [description] => https://www.php.net/
    //     [keywords] =>
    //     [icons] => Array
    //         (
    //             [0] => https://www.php.net/favicon.ico
    //         )

    // )

    $uriMeta = $extracter->extract('https://github.com/php/php-src');

    print_r($uriMeta->toArray());
    // Array
    // (
    //     [uri] => https://github.com/php/php-src
    //     [host] => github.com
    //     [scheme] => https
    //     [title] => GitHub - php/php-src: The PHP Interpreter
    //     [description] => The PHP Interpreter. Contribute to php/php-src development by creating an account on GitHub.
    //     [keywords] =>
    //     [icons] => Array
    //         (
    //             [0] => https://github.githubassets.com/favicons/favicon.svg
    //             [1] => https://github.githubassets.com/pinned-octocat.svg
    //             [2] => https://github.com/favicon.ico
    //         )
    // )
```

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `uri-meta` will be documented in this file.

## 1.0.0 - 202X-XX-XX

-   initial release


tests/BaseTest.php
<?php

namespace Chuoke\UriMeta\Tests;

use Chuoke\UriMeta\UriMetaExtracter;
use PHPUnit\Framework\TestCase;

class BaseTest extends TestCase
{
    /** @test */
    public function can_use()
    {
        $extracter = new UriMetaExtracter();


        $uriMeta = $extracter->extract('https://www.baidu.com/');

        $this->assertEquals('百度一下，你就知道', $uriMeta->title());


        $uriMeta = $extracter->extract('https://www.php.net/');

        $this->assertEquals('PHP: Hypertext Preprocessor', $uriMeta->title());


        $uriMeta = $extracter->extract('https://github.com/php/php-src');

        $this->assertEquals('GitHub - php/php-src: The PHP Interpreter', $uriMeta->title());
    }
}


src/UriMeta.php
<?php

namespace Chuoke\UriMeta;

use League\Uri\Uri;

class UriMeta
{
    /** @var \League\Uri\Uri */
    protected $uri;

    public array $titles = [];

    public array $descriptions = [];

    public string $keywords = '';

    public array $icons = [];

    public function __construct($uri)
    {
        if (! $uri instanceof Uri) {
            $uri = Uri::createFromString($uri);
        }

        $this->uri = $uri;
    }

    public function uri(): string
    {
        return (string) $this->uri;
    }

    public function host(): ?string
    {
        return $this->uri->getHost();
    }

    public function scheme(): ?string
    {
        return $this->uri->getScheme();
    }

    public function title(): string
    {
        return reset($this->titles) ?: $this->defaultTitle();
    }

    public function defaultTitle(): string
    {
        return (string) $this->host();
    }

    public function titles(): array
    {
        return $this->titles;
    }

    public function description(): string
    {
        return (string) reset($this->descriptions);
    }

    public function descriptions(): array
    {
        return $this->descriptions;
    }

    public function keywords(): string
    {
        return $this->keywords;
    }

    public function icon(): string
    {
        return (string) reset($this->icons);
    }

    public function icons(): array
    {
        return $this->icons;
    }

    public function toArray(): array
    {
        return [
            'uri' => $this->uri(),
            'host' => $this->host(),
            'scheme' => $this->scheme(),
            'title' => $this->title(),
            'description' => $this->description(),
            'keywords' => $this->keywords(),
            'icons' => $this->icons(),
        ];
    }
}


config/urimeta.php
<?php

return [

    'user_agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',

    'possibles' => [
        'title' => [
            'meta[name=apple-mobile-web-app-title]' => 'content',
            'meta[name=og:site_name]' => 'content',
            'meta[name=application-name]' => 'content',
        ],
        'sub_title' => [
            'meta[name=twitter:title]' => 'content',
            'meta[property=og:title]' => 'content',
            'meta[name=title]' => 'content',
            'title' => 'text',
        ],
        'description' => [
            'meta[name=twitter:description]' => 'content',
            'meta[property=og:description]' => 'content',
            'meta[name=description]' => 'content',
            'meta[name=Description]' => 'content',
        ],
        'keywords' => [
            'meta[name=keywords]' => 'content',
            'meta[name=Keywords]' => 'content',
        ],
        'icon' => [
            'link[rel="apple-touch-icon"]' => 'href',
            'link[rel="icon"]' => 'href',
            'link[rel="shortcut icon"]' => 'href',
            'link[rel="Shortcut Icon"]' => 'href',
            'link[rel="mask-icon"]' => 'href',
        ],
    ],

    /*
    | URIs that start with the following strings will not be resolved.
    */
    'skips' => [
        'localhost',
        '127',
        '192',
    ],
];


src/UriMetaExtracter.php
<?php

namespace Chuoke\UriMeta;

use DiDom\Document;
use Exception;
use HeadlessChromium\BrowserFactory;
use League\Uri\Uri;

class UriMetaExtracter
{
    protected array $config;

    /** @var \League\Uri\Uri */
    protected $uri;

    /** @var \DiDom\Document */
    protected $document;

    public function __construct(array $config = null)
    {
        $this->setConfig($config ?: []);
    }

    /**
     * Merge the given configuration with the existing configuration.
     *
     * @param  array  $config
     * @return $this
     */
    public function setConfig(array $config)
    {
        $this->config = array_merge(
            require __DIR__ . '/../config/urimeta.php',
            $config
        );

        return $this;
    }

    /**
     * @param  string  $uri
     * @return \Chuoke\UriMeta\UriMeta
     * @throws Exception
     */
    public function extract(string $uri)
    {
        $this->uri = Uri::createFromString($uri);

        if (! $this->isHttp() || $this->isShouldSkip()) {
            return $this->useDefault();
        }

        return $this->extractMeta();
    }

    protected function isHttp(): bool
    {
        return stripos($this->uri->getScheme(), 'http') === 0;
    }

    protected function isShouldSkip(): bool
    {
        $host = $this->uri->getHost();

        foreach ($this->config['skips'] ?? [] as $possible) {
            if (\str_starts_with($host, $possible)) {
                return true;
            }
        }

        return false;
    }

    /**
     * @return \Chuoke\UriMeta\UriMeta
     */
    public function useDefault()
    {
        $uriMeta = new UriMeta($this->uri);

        $uriMeta->titles = [$this->defaultTitle()];
        $uriMeta->descriptions = [(string) $this->uri];
        $uriMeta->keywords = '';
        $uriMeta->icons = [];

        return $uriMeta;
    }

    /**
     * @return \Chuoke\UriMeta\UriMeta
     */
    protected function extractMeta()
    {
        $this->makeDocument();

        $uriMeta = new UriMeta($this->uri);

        $uriMeta->titles = $this->extractTitle();
        $uriMeta->descriptions = $this->extractDescription();
        $uriMeta->keywords = $this->extractKeywords();
        $uriMeta->icons = $this->extractIcons();

        return $uriMeta;
    }

    public function defaultTitle()
    {
        return implode(
            ':',
            array_filter([
                $this->uri->getHost(),
                $this->uri->getPort(),
            ])
        );
    }

    /**
     * 提取meta标签值
     *
     * @param  array  $possibles
     * @param  bool  $takeAll
     * @return array
     */
    protected function extractMetaValue(array $possibles, $takeAll = false)
    {
        if (! ($headEle = $this->document->first('head'))) {
            return [];
        }

        $values = [];

        foreach ($possibles as $possible => $attr) {
            if (! ($possibleEle = $headEle->first($possible))) {
                continue;
            }

            $value = trim(
                strcasecmp($attr, 'text') === 0
                    ? $possibleEle->text()
                    : $possibleEle->getAttribute($attr)
            );

            if ($value) {
                $values[] = $value;
            }

            if (! empty($values) && ! $takeAll) {
                break;
            }
        }

        return array_unique($values);
    }

    protected function extractTitle()
    {
        $possibles = $this->config['possibles']['sub_title'];

        if ($this->isHostUri()) {
            $possibles = array_merge($this->config['possibles']['title'], $possibles);
        }

        return $this->extractMetaValue($possibles, true);
    }

    protected function extractDescription()
    {
        $possibles = $this->config['possibles']['description'];

        return $this->extractMetaValue($possibles, true);
    }

    protected function extractKeywords()
    {
        $possibles = $this->config['possibles']['keywords'];

        return $this->extractMetaValue($possibles, false)[0] ?? '';
    }

    protected function extractIcons()
    {
        $possibles = $this->config['possibles']['icon'];

        $icons = $this->extractMetaValue($possibles, true);

        $icons[] = '/favicon.ico';

        $results = [];

        foreach (array_unique($icons) as $icon) {
            $results[] = $this->fulfillSubUrl($icon);
        }

        return array_unique($results);
    }

    public function fulfillSubUrl($sub)
    {
        if ($this->startsWith($sub, ['https://', 'http://'])) {
            return $sub;
        } elseif ($this->startsWith($sub, '//')) {
            return implode(':', [$this->uri->getScheme() ?: 'http', $sub]);
        } elseif ($this->startsWith($sub, '/')) {
            return $this->hostUri() . $sub;
        }

        return implode('/', array_filter([
            $this->hostUri(),
            trim($this->uri->getPath(), '/'),
            $sub,
        ]));
    }

    protected function isHostUri(): bool
    {
        return ! $this->uri->getPath();
    }

    public function hostUri(): string
    {
        return implode('', [
            $this->uri->getScheme() ?: 'http',
            '://',
            $this->uri->getHost(),
        ]);
    }

    protected function makeDocument()
    {
        if (strpos($this->uri->getScheme(), 'http') !== 0) {
            throw new Exception('Non-web link');
        }

        $html = $this->getHtmlbyChrome();

        return $this->document = new Document($html, false, 'UTF-8');
    }

    public function getHtmlbyChrome()
    {
        try {
            $browserFactory = new BrowserFactory();

            // starts headless chrome
            $browser = $browserFactory->createBrowser();

            // creates a new page and navigate to an url
            $page = $browser->createPage();

            if ($userAgent = $this->userAgent()) {
                $page->setUserAgent($userAgent);
            }

            $page->navigate((string) $this->uri)->waitForNavigation();

            return $page->getHtml();
        } finally {
            isset($browser) && $browser && $browser->close();
        }
    }

    /**
     * Determine if a given string starts with a given substring.
     *
     * @param  string  $haystack
     * @param  string|string[]  $needles
     * @return bool
     */
    protected function startsWith($haystack, $needles)
    {
        foreach ((array) $needles as $needle) {
            if (str_starts_with($haystack, $needle)) {
                return true;
            }
        }

        return false;
    }

    protected function userAgent()
    {
        return $this->config['user_agent'] ?? 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36';
    }
}
